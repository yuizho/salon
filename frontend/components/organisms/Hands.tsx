import { FC } from 'react';
import { useRecoilState, useRecoilValue, useSetRecoilState } from 'recoil';
import * as Sentry from '@sentry/nextjs';
import pick from '../../graphql/clients/pick';
import { NETWORK_ERROR } from '../../graphql/error-message';
import { Status } from '../../graphql/schema';
import { appState } from '../../states/app';
import { myState } from '../../states/me';
import { pokerState } from '../../states/poker';
import { usersState } from '../../states/users';
import Card from '../atoms/Card';

type Props = {
  values: Array<string>;
};

type ComponentProps = Props & {
  glow: boolean;
  choosable: boolean;
  onClick: (pickedCard: string) => Promise<boolean>;
};

export const Component: FC<ComponentProps> = ({ values, glow, choosable, onClick }) => (
  <div className="flex flex-wrap justify-center gap-2">
    {values.map((v) => (
      <Card
        key={v}
        value={v}
        shown
        choosable={choosable}
        onClick={onClick}
        chosen={false}
        glow={glow}
      />
    ))}
  </div>
);

const Container: FC<Props> = ({ values }) => {
  const [me] = useRecoilState(myState);
  const setUsers = useSetRecoilState(usersState);
  const poker = useRecoilValue(pokerState);
  const setApp = useSetRecoilState(appState);

  const onClickCard = (roomId: string, userId: string) => async (pickedCard: string) => {
    setApp((app) => ({ ...app, loading: true }));

    // update user state before sending request to GraphQL API
    // to show the result of user operation as soon as possible.
    setUsers((currentUsers) => {
      const users = [
        ...currentUsers.filter((u) => u.userId !== userId),
        {
          userId,
          status: Status.CHOSEN,
          pickedCard,
        },
      ];
      users.sort((a, b) => (a.userId > b.userId ? 1 : -1));
      return users;
    });

    try {
      const result = await pick(roomId, userId, me.userToken, pickedCard);
      return result.data?.pick.user_id === userId ?? false;
    } catch (e) {
      Sentry.captureException(e);
      setApp((app) => ({
        ...app,
        loading: false,
        errorMessage: NETWORK_ERROR,
      }));
      return false;
    }
  };

  return (
    <Component
      values={values}
      glow={poker.state === 'CHOOSING'}
      choosable={poker.state !== 'KICKED'}
      onClick={onClickCard(me.roomId, me.userId)}
    />
  );
};

export default Container;
