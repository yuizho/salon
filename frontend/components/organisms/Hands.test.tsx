import { render } from '@testing-library/react';
import { RecoilRoot } from 'recoil';

import Hands from './Hands';

describe('Counter', () => {
  test('render', () => {
    const { asFragment } = render(
      <RecoilRoot>
        <Hands values={['1', '2', '3']} />
      </RecoilRoot>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
