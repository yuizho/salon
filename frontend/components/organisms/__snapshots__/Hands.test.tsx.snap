// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hands snapshot 1`] = `
<DocumentFragment>
  <div
    class="
  flex flex-wrap gap-1
"
  >
    <div
      class="relative"
    >
      <div
        class="relative"
      >
        <div
          class="
      
flex
items-center justify-center
w-16 h-20
border-slate-400
rounded border
text-slate-600
select-none
bg-white

      
shadow-md
hover:cursor-pointer
hover:border-2
active:shadow-none
active:border
active:border-slate-300

      
      font-bold text-2xl
    "
        >
          1
        </div>
      </div>
    </div>
    <div
      class="relative"
    >
      <div
        class="relative"
      >
        <div
          class="
      
flex
items-center justify-center
w-16 h-20
border-slate-400
rounded border
text-slate-600
select-none
bg-white

      
shadow-md
hover:cursor-pointer
hover:border-2
active:shadow-none
active:border
active:border-slate-300

      
      font-bold text-2xl
    "
        >
          2
        </div>
      </div>
    </div>
    <div
      class="relative"
    >
      <div
        class="relative"
      >
        <div
          class="
      
flex
items-center justify-center
w-16 h-20
border-slate-400
rounded border
text-slate-600
select-none
bg-white

      
shadow-md
hover:cursor-pointer
hover:border-2
active:shadow-none
active:border
active:border-slate-300

      
      font-bold text-2xl
    "
        >
          3
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
